services/block_service.go

```diff
@@ -105,18 +105,33 @@ func (s *BlockAPIService) PopulateTransaction(
 		receiptLogs = tx.Receipt.Logs
 	}
 	// Compute tx operations via tx.Receipt logs for ERC20 transfer, mint and burn
+	// Declare a map to act as the cache
+	var contractCurrencyCache = make(map[string]*evmClient.ContractCurrency)
+
 	for _, log := range receiptLogs {
-		// if Filter == false, we record every ERC20 tokens
-		if !s.client.GetRosettaConfig().FilterTokens || (s.client.GetRosettaConfig().FilterTokens &&
-			client.IsValidERC20Token(s.client.GetRosettaConfig().TokenWhiteList, log.Address.String())) {
+		if !s.client.GetRosettaConfig().FilterTokens ||
+			(s.client.GetRosettaConfig().FilterTokens &&
+				evmClient.IsValidERC20Token(s.client.GetRosettaConfig().TokenWhiteList, log.Address.String())) {
+
 			switch len(log.Topics) {
 			case TopicsInErc20DepositOrWithdrawal, TopicsInErc20Transfer:
-				currency, err := s.client.GetContractCurrency(log.Address, true)
-				if err != nil {
-					return nil, err
+				addressStr := log.Address.String()
+				var currency *evmClient.ContractCurrency
+				var err error
+
+				// Check if value exists in cache
+				if cachedCurrency, found := contractCurrencyCache[addressStr]; found {
+					currency = cachedCurrency
+				} else {
+					// If not in cache, call GetContractCurrency and store the result in cache
+					currency, err = s.client.GetContractCurrency(log.Address, true)
```

Comment:
all contracts share the same ABI, if the contract address is new, yes we need the ABI call under GetContractCurrency call then cache the currency, but if the contract address is old (i.e. it's already cached), we can leverage the cache and we don't need GetContractCurrency call (i.e. ABI call)

Reply:
when `PopulateTransaction` is being used in /block service, it will try to populate all the transactions in a block (not only all the internal txs inside a normal tx). As a result, contract address could be different among different transactions inside a block. The cache here is basically a map to store the contract address -> contract currency mapping. So that in runtime we could check map first before initiate a client call